@ladvoc looks like it won't work:

```
error[E0425]: cannot find type `UniFfiTag` in crate `bytes`

The first path segment isn't the crate that defines the type — it's the crate whose registration you're borrowing. The macro expands to <Bytes as FfiConverter<bytes::UniFfiTag>>, and bytes isn't a UniFFI crate, so it has no UniFfiTag. The docs are genuinely misleading here: their example borrows from another uniffi crate, so the path reads like a plain type path. You need a crate that called setup_scaffolding!().
```

--- a/.changeset/reuse_datatrack_bytes_converter.md
+++ b/.changeset/reuse_datatrack_bytes_converter.md
@@ -1,5 +1,7 @@
 ---
 livekit-uniffi: patch
+livekit-datatrack: patch
+livekit-common: patch
 ---
 
-Reuse `livekit-datatrack`'s `Bytes` converter instead of registering a second one, so UniFFI emits it once and the post-generation Swift workaround is no longer needed
+Register the `Bytes` UniFFI custom type once in `livekit-common` and borrow it from each component with `uniffi::use_remote_type!`, so the converter is emitted once and the post-generation Swift workaround is no longer needed

--- a/.github/workflows/uniffi-swift.yml
+++ b/.github/workflows/uniffi-swift.yml
@@ -85,7 +85,8 @@ jobs:
 
       - name: Publish to hosting repo
         if: ${{ !inputs.dry_run }}
-        uses: livekit/publish-xcframework-action@200ee4984ef1b68068c0b3c4e87e798988730c90 # main @ fix: abort on missing source files (#3)
+        # TODO: repin to the merge commit once livekit/publish-xcframework-action#5 lands.
+        uses: livekit/publish-xcframework-action@94cb0ea28a25e18469c4a3e43c7f162a84da2169 # feat/glob-sources @ feat: allow path patterns as file sources (#5)
         with:
           xcframework-zip: ${{ env.OUTPUT_DIR }}/Rust${{ env.SPM_NAME }}.xcframework.zip
           version: ${{ inputs.version }}
@@ -96,12 +97,9 @@ jobs:
             ${{ env.OUTPUT_DIR }}/${{ env.SPM_NAME }}.podspec:${{ env.SPM_NAME }}.podspec
             ${{ env.OUTPUT_DIR }}/LICENSE:LICENSE
             ${{ env.OUTPUT_DIR }}/PrivacyInfo.xcprivacy:PrivacyInfo.xcprivacy
-            # One entry per UniFFI component (each `setup_scaffolding!()` crate gets its own
-            # generated file). Add a line here when a component is added, or the published
-            # package will be missing types and fail to build for consumers.
-            ${{ env.OUTPUT_DIR }}/Sources/${{ env.SPM_NAME }}/livekit_uniffi.swift:Sources/${{ env.SPM_NAME }}/livekit_uniffi.swift
-            ${{ env.OUTPUT_DIR }}/Sources/${{ env.SPM_NAME }}/livekit_common.swift:Sources/${{ env.SPM_NAME }}/livekit_common.swift
-            ${{ env.OUTPUT_DIR }}/Sources/${{ env.SPM_NAME }}/livekit_datatrack.swift:Sources/${{ env.SPM_NAME }}/livekit_datatrack.swift
+            # UniFFI emits one source per component, so match rather than list them:
+            # adding a component must not require a workflow change.
+            ${{ env.OUTPUT_DIR }}/Sources/${{ env.SPM_NAME }}/*.swift:Sources/${{ env.SPM_NAME }}/
           token: ${{ secrets.UNIFFI_XCFRAMEWORK_PAT }}
           pr-body: |
             Source tag: `${{ inputs.tag_name }}`

--- a/.github/workflows/uniffi-swift.yml
+++ b/.github/workflows/uniffi-swift.yml
@@ -96,7 +96,11 @@ jobs:
             ${{ env.OUTPUT_DIR }}/${{ env.SPM_NAME }}.podspec:${{ env.SPM_NAME }}.podspec
             ${{ env.OUTPUT_DIR }}/LICENSE:LICENSE
             ${{ env.OUTPUT_DIR }}/PrivacyInfo.xcprivacy:PrivacyInfo.xcprivacy
+            # One entry per UniFFI component (each `setup_scaffolding!()` crate gets its own
+            # generated file). Add a line here when a component is added, or the published
+            # package will be missing types and fail to build for consumers.
             ${{ env.OUTPUT_DIR }}/Sources/${{ env.SPM_NAME }}/livekit_uniffi.swift:Sources/${{ env.SPM_NAME }}/livekit_uniffi.swift
+            ${{ env.OUTPUT_DIR }}/Sources/${{ env.SPM_NAME }}/livekit_common.swift:Sources/${{ env.SPM_NAME }}/livekit_common.swift
             ${{ env.OUTPUT_DIR }}/Sources/${{ env.SPM_NAME }}/livekit_datatrack.swift:Sources/${{ env.SPM_NAME }}/livekit_datatrack.swift
           token: ${{ secrets.UNIFFI_XCFRAMEWORK_PAT }}
           pr-body: |

--- a/livekit-common/Cargo.toml
+++ b/livekit-common/Cargo.toml
@@ -9,10 +9,10 @@ repository.workspace = true
 
 [dependencies]
 livekit-protocol = { workspace = true }
-bytes = { workspace = true }
+bytes = { workspace = true, optional = true }
 uniffi = { workspace = true, features = ["scaffolding-ffi-buffer-fns"], optional = true }
 
 [features]
 # Exposes this crate's shared FFI type registrations. Enabled transitively by
 # livekit-uniffi so every component borrows one converter per type.
-uniffi = ["dep:uniffi"]
+uniffi = ["dep:uniffi", "dep:bytes"]

--- a/livekit-common/Cargo.toml
+++ b/livekit-common/Cargo.toml
@@ -9,3 +9,10 @@ repository.workspace = true
 
 [dependencies]
 livekit-protocol = { workspace = true }
+bytes = { workspace = true }
+uniffi = { workspace = true, features = ["scaffolding-ffi-buffer-fns"], optional = true }
+
+[features]
+# Exposes this crate's shared FFI type registrations. Enabled transitively by
+# livekit-uniffi so every component borrows one converter per type.
+uniffi = ["dep:uniffi"]

--- a/livekit-common/src/ffi_types.rs
+++ b/livekit-common/src/ffi_types.rs
@@ -1,0 +1,29 @@
+// Copyright 2026 LiveKit, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Shared UniFFI type registrations.
+//!
+//! A remote type such as [`bytes::Bytes`] can only be registered once per UniFFI component:
+//! `custom_type!` emits a `public` converter per component, and the generated Swift files are
+//! compiled into a single module, so a second registration fails to build. Registering it here
+//! and having each component borrow it with `uniffi::use_remote_type!` keeps exactly one
+//! declaration no matter how many crates need the type.
+//!
+//! Owning a registration requires being a UniFFI component: `custom_type!` needs
+//! `crate::UniFfiTag`, and bindgen rejects type metadata that belongs to no namespace
+//! (`Unknown namespace for CustomType`). Hence the `setup_scaffolding!` in `lib.rs`.
+
+use bytes::Bytes;
+
+uniffi::custom_type!(Bytes, Vec<u8>, { remote });

--- a/livekit-common/src/lib.rs
+++ b/livekit-common/src/lib.rs
@@ -22,6 +22,14 @@ use livekit_protocol as proto;
 
 mod enum_dispatch;
 
+// Must sit at the crate root: it defines `crate::UniFfiTag`, which the registrations in
+// `ffi_types` resolve against.
+#[cfg(feature = "uniffi")]
+uniffi::setup_scaffolding!();
+
+#[cfg(feature = "uniffi")]
+mod ffi_types;
+
 // -------------------------------------------------------------------------------------------------
 // Client protocol
 // -------------------------------------------------------------------------------------------------

--- a/livekit-common/uniffi.toml
+++ b/livekit-common/uniffi.toml
@@ -1,0 +1,4 @@
+[bindings.swift]
+# Matches the Rust<Name> convention used by the other components; the cargo-swift
+# fork folds every component's header into the single RustLiveKitUniFFI framework.
+ffi_module_name = "RustLiveKitCommon"

--- a/livekit-datatrack/Cargo.toml
+++ b/livekit-datatrack/Cargo.toml
@@ -9,6 +9,7 @@ repository.workspace = true
 
 [dependencies]
 livekit-protocol = { workspace = true }
+livekit-common = { workspace = true, optional = true }
 livekit-runtime = { workspace = true, features = ["tokio"] }
 log = { workspace = true }
 thiserror = { workspace = true }
@@ -25,7 +26,7 @@ uniffi = { workspace = true, features = ["scaffolding-ffi-buffer-fns"], optional
 indexmap = "2"
 
 [features]
-uniffi = ["dep:uniffi"]
+uniffi = ["dep:uniffi", "dep:livekit-common", "livekit-common/uniffi"]
 __fuzz = ["dep:fake"]
 
 [dev-dependencies]

--- a/livekit-datatrack/src/e2ee.rs
+++ b/livekit-datatrack/src/e2ee.rs
@@ -76,7 +76,7 @@ pub trait DecryptionProvider: Send + Sync + Debug {
 }
 
 #[cfg(feature = "uniffi")]
-uniffi::custom_type!(Bytes, Vec<u8>, { remote });
+uniffi::use_remote_type!(livekit_common::Bytes);
 
 #[cfg(feature = "uniffi")]
 uniffi::custom_type!(InitializationVector, Vec<u8>, {

--- a/livekit-uniffi/Cargo.toml
+++ b/livekit-uniffi/Cargo.toml
@@ -13,6 +13,7 @@ publish = false
 
 [dependencies]
 livekit-protocol = { workspace = true }
+livekit-common = { workspace = true, features = ["uniffi"] }
 livekit-api = { workspace = true, default-features = false, features = ["access-token"] }
 livekit-datatrack = { workspace = true, features = ["uniffi"] }
 uniffi = { workspace = true, features = ["scaffolding-ffi-buffer-fns", "tokio"] }

--- a/livekit-uniffi/src/common.rs
+++ b/livekit-uniffi/src/common.rs
@@ -14,10 +14,8 @@
 
 use bytes::Bytes;
 
-// `Bytes` is a remote type, so each UniFFI component that registers it with
-// `custom_type!` emits its own converter — and the two component Swift files are
-// compiled into one module, so a second `typealias Bytes` fails to build
-// ("invalid redeclaration of 'Bytes'"). Reuse livekit-datatrack's registration
-// instead of adding a second one; the type is then emitted once, in the file that
-// owns it. Upstream: https://github.com/mozilla/uniffi-rs/issues/2933
-uniffi::use_remote_type!(livekit_datatrack::Bytes);
+// `Bytes` is registered once in livekit-common so every component that needs it borrows the
+// same converter; registering it here as well would emit a second `public typealias Bytes`
+// into this component's Swift file, and both files compile into one module.
+// Upstream: https://github.com/mozilla/uniffi-rs/issues/2933
+uniffi::use_remote_type!(livekit_common::Bytes);

